receiver/prometheusexecreceiver/receiver.go

```diff
@@ -0,0 +1,313 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheusexecreceiver
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/config"
+	sdconfig "github.com/prometheus/prometheus/discovery/config"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/receiver/prometheusreceiver"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusexecreceiver/subprocessmanager"
+)
+
+const (
+	minPortRange       int           = 10000            // Minimum of the port generation range
+	maxPortRange       int           = 11000            // Maximum of the port generation range
+	portTemplate       string        = "{{port}}"       // Template for port in strings
+	healthyProcessTime time.Duration = 30 * time.Minute // healthyProcessTime is the default time a process needs to stay alive to be considered healthy
+	healthyCrashCount  int           = 3                // healthyCrashCount is the amount of times a process can crash (within the healthyProcessTime) before being considered unstable - it may be trying to find a port
+
+)
+
+// Local random seed to not override anything being used globally
+var random *rand.Rand = rand.New(rand.NewSource(time.Now().UnixNano()))
+
+type prometheusReceiverWrapper struct {
+	logger   *zap.Logger
+	config   *Config
+	consumer consumer.MetricsConsumerOld
+
+	// Prometheus receiver config
+	receiverConfig *prometheusreceiver.Config
+
+	// Subprocess data
+	subprocessConfig *subprocessmanager.SubprocessConfig
+
+	// Receiver data
+	originalPort       int
```

Comment:
Why `originalPort` rather than just `port`?

Reply:
I found it's better to call it original port since the actual port the receiver/process pair may be using at any given time has an actual value while the original port is 0 when not defined by the user. makes it clearer in my conditions what i'm comparing!